app/src/test/java/org/sagebionetworks/bridge/researchstack/CrfDataProviderTest.java

```diff
@@ -0,0 +1,514 @@
+    }
+
+    @Test
+    public void testFindClinc1() {
+        final ScheduledActivityListV4 successList = successList(true);
+        final ScheduledActivity clinic = dataProvider.findActivity(successList, "clinic1");
+        assertNotNull(clinic);
+        assertEquals("clinic1", clinic.getActivity().getTask().getIdentifier());
+    }
+
+    @Test
+    public void testFindClinc2() {
+        final ScheduledActivityListV4 successList = successList(true);
+        final ScheduledActivity clinic = dataProvider.findActivity(successList, "clinic2");
+        assertNotNull(clinic);
+        assertEquals("clinic2", clinic.getActivity().getTask().getIdentifier());
+    }
+
+    @Test
+    public void testFindFail() {
+        final ScheduledActivityListV4 successList = successList(true);
+        final ScheduledActivity clinic = dataProvider.findActivity(successList, "clinic3");
+        assertNull(clinic);
+    }
+
+    @Test
+    public void testAssignClinic1() {
+        resetTest(null, true, null);
+        final ScheduledActivityListV4 successList = successList(true);
+        dataProvider.mockGetAllActiities(successList);
+
+        dataProvider.mockGetClinicActiities(clinicActivities(null, null));
+        dataProvider.getCrfActivities(new CrfDataProvider.CrfActivitiesListener() {
+            @Override
+            public void success(ScheduledActivityListV4 activityList) {
+                assertEquals(activityList, successList);
+                assertEquals(dataProvider.getCrfPrefs().getFirstSignInDate(), dataProvider.COMPLETION_TIME_CLINIC1);
+            }
+
+            @Override
+            public void error(String localizedError) {
+                assertFalse(true);  // fail
+            }
+        });
+    }
+
+    @Test
+    public void testAssignClinic2() {
+        resetTest(null, false, null);
+        final ScheduledActivityListV4 successList = successList(false);
+        dataProvider.mockGetAllActiities(successList);
+
+        dataProvider.mockGetClinicActiities(clinicActivities(null, null));
+        dataProvider.getCrfActivities(new CrfDataProvider.CrfActivitiesListener() {
+            @Override
+            public void success(ScheduledActivityListV4 activityList) {
+                assertEquals(activityList, successList);
+                assertEquals(dataProvider.getCrfPrefs().getFirstSignInDate(), dataProvider.COMPLETION_TIME_CLINIC2);
+            }
+
+            @Override
+            public void error(String localizedError) {
+                assertFalse(true);  // fail
+            }
+        });
+    }
+
+    @Test
+    public void testExistingServerDataGroupClinic1() {
+        resetTest(null, true, "clinic1");
+        final ScheduledActivityListV4 successList = successList(true);
+        dataProvider.mockGetAllActiities(successList);
+
+        dataProvider.mockGetClinicActiities(clinicActivities(null, null));
+        dataProvider.getCrfActivities(new CrfDataProvider.CrfActivitiesListener() {
```

Comment:
Similarly, should these tests update that we correctly update the clinic activity?